activesupport/lib/active_support/duration/iso8601_parser.rb

```diff
@@ -0,0 +1,72 @@
+require 'active_support/core_ext/hash/slice'
+
+module ActiveSupport
+  class Duration
+    # Parses a string formatted according to ISO 8601 Duration into the hash .
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # Parts of code are taken from ISO8601 gem by Arnau Siches (@arnau).
+    # This parser isn't so strict and allows negative parts to be present in pattern.
+    class ISO8601Parser
+      attr_reader :parts
+
+      class ParsingError < ::StandardError; end
+
+      def initialize(iso8601duration)
+        @raw = iso8601duration
+        parse_iso_duration!
+        construct_parts
+        validate_parts!
+      end
+
+      private
+        def parse_iso_duration!
+          @match = @raw.match(/^
+                      (?<sign>\+|-)?
```

Comment:
And what about individual components being negative, e.g. `P1M-1D` - have you seen any of those in the wild?